### ISSUE TITLE
backend: Add Apple iCloud Photos backend

### ISSUE THREAD

**Issue by stv0g** [rclone/rclone#7982]
#### What is your current rclone version (output from `rclone version`)?

v1.67

#### What problem are you are trying to solve?

Apple's iCloud can store both files in the iCloud Drive and photos via Apples Photos app.
A new backend for accessing the iCloud Drive is currently beeing finalized in https://github.com/rclone/rclone/issues/1778 & https://github.com/rclone/rclone/pull/7717

This issues attempts to track the development of extending the initial iCloud Drive support to photos.

#### How do you think rclone should be changed to solve that?

The existing iCloud Drive backend should be extended with support for accessing photos.
@lostb1t, the author of the initial iCloud backend implementation, has outlined the planned approach [in this comment](https://github.com/rclone/rclone/issues/1778#issuecomment-2029300270):

> > B: single remote includes all available services mapped to their names in the root(so icloudremote:/drive, icloudremote:/photos etc)
> > 
> > C: B but let user choose woch services are avaiable for the remote
> > 
> > A and C are probably quite similar just with a config option to control. One of those sounds the easiest to implement and there is lots of precedent in rclone, eg drive/shared drives or OneDrive/SharePoint
> > 
> 
> im gonna focus on option C. This allows credentials to be defined only once and a user can choose the services from the config.
> 
> **Il write an Fs interface for every service but combine them similar to combine backen to provide a single backend to rclone**

So the plan would be to combine both Drive and Photos in a single backend but use two separate `Fs` implementations internally.

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by stv0g** [comment]
I would be willing to sponsor this feature with a contribution of 50 USD.

@Scripter17 also offered to extend his funding to 200 USD (100 USD for Photos support) in [this comment](https://github.com/rclone/rclone/pull/7717#issuecomment-2261025756).

**Comment by mlkt** [comment]
Thanks to very much for this great project. It is a very important feature for me to back up icloud photos with rclone. I hope rclone can realize the function of listing photos by album (including hidden photos).

**Comment by xz-dev** [comment]
FYI, on my test, iCloud and iCloud Photo seems totally different things.
Because even if you use Mac, move photo from iCloud to iCloud Photo, it will also download from iCloud and then upload to iCloud Photo.

**Comment by hasnat** [comment]
you're right @xz-dev ; they are different. they are likely different sort of different zones `com.apple.CloudDocs` vs am unsure whats the handle for Photos zone/db . data very likely would be organized different to iCloud Drive.

**Comment by lifenautjoe** [comment]
> I am willing to sponsor this feature with a contribution of 50 USD.
> 
> [@Scripter17](https://github.com/Scripter17) also offered to extend his funding to 200 USD (100 USD for Photos support) in [this comment](https://github.com/rclone/rclone/pull/7717#issuecomment-2261025756).

I'd like to extend the funding by $500 USD, ($700 in total)

**Comment by poolpOrg** [comment]
I'll extend the funding by $1000 :-)

**Comment by stv0g** [comment]
Lets make this pledges a bit more binding. I've added this issue to GitPay, a bounty collection provider:

https://gitpay.me/#/task/1113

Bounties will be distributed once this ticket has been closed and a PR has been merged.

**Comment by lostb1t** [comment]
🤔

**Comment by Scripter17** [comment]
I'm wary of putting my bank details and $100 into a website I've never heard of before, so I'm going to just rely on my reputation for this. I make good on my bounties and several other people can attest to that

**Comment by Scripter17** [comment]
> Hello @Scripter17 , I'm the Gitpay maintainer. We have the platform 5+ already paying 20k+ to contributors, and we use Stripe for payments. It is secure, and we are fully open source. Gitpay helps to facilitate payments anywhere in the world.
> 
> I'm a developer, too, and we support open source and help developers contribute.
> 
> We also have companies and open-source projects that have been used for a while. Let me know if you have any questions.

Unfortunately I can't just take your word for that; It's reasonable to assume you're at least working in good faith because why else would it be open source, but I can't take the word of people directly involved in the project

I'll be sure to have a look around to form a chainmail of trust, because I do agree that binding people to their bounties is important, but for me, for now, I think I'll rely on my reputation (especially since I've already made good on a $100 in rclone)

~~Also it being written in javascript does not fill me with confidence but that's purely prejudice. #Rustgang~~

**Comment by lostb1t** [comment]
@Myestery if nobody else
picks it up i might

**Comment by lostb1t** [comment]
go for it

**Comment by alexanmtz** [comment]
And another thing. @Scripter17 , even if you have a reputation and you will give the bounty for sure, I guess it's better for the developers working on it to have all payments from other sponsors in one go, instead of picking every bounty individually for every person who contributes.

**Comment by Sameerali0** [comment]
@Scripter17 I can vouch for @alexanmtz as I have recieved bounties from gitpay platform and It would be better for the bounty hunter if all of the payment is in one place and he is satisfied that once the work is completed he will recieve the payment

**Comment by alexanmtz** [comment]
@Scripter17, the issue was posted on Gitpay, and our users are aware of it. They are looking for work on issues, so the idea behind the platform is also to attract more contributors willing to work on solving issues.

**Comment by Scripter17** [comment]
> @Scripter17, the issue was posted on Gitpay, and our users are aware of it. They are looking for work on issues, so the idea behind the platform is also to attract more contributors willing to work on solving issues. 

Not sure why I didn't realize that. Sorry

**Comment by poolpOrg** [comment]
I'm also on the fence for using gitpay at the moment as I had not known about the service until now, not opposed to using it but clearly I won't put 1000 there without doing a bit of research first.

Anyone knows what happens if no one works on this, do gitpay keep it indefinitely ?

**Comment by alexanmtz** [comment]
@poolpOrg, yes, it can stay there. If it's not solved or closed, some bounties take a while, while others don't. Alternatively, you can get a full refund on the platform if no one claims it.

**Comment by suraj-mandal** [comment]
@alexanmtz @poolpOrg  If no one is currently working, I would like to try this.

**Comment by jayantpranjal0** [comment]
@stv0g I would like to wokr on this issue. I would first open a design a document or a discussion thread on how to approach this issue and then work on this. Please let me know if it would be okayy for me to work on this.

**Comment by stv0g** [comment]
As in the previous comment: I am not aware of anybody working on this yet.
@lostb1t implemented the support for iCloud files. I think it might be wise to coordinate with him and reuse as much as possible from his implementation for iCloud photos.

**Comment by coughlanio** [comment]
Hey all 👋

I've been working on implementing an iCloud Photos backend for rclone and have made good progress on several key areas:

- Authentication and session handling with iCloud services (leveraging some of @lostb1t's iclouddrive backend code 🎉)
- Core directory/download functionality working and tested
- Support for both PrimarySync and SharedSync libraries
- Read-only backend implementation following rclone conventions

The implementation is showing promising results across different testing scenarios (120k+ photos), including various album types and media formats.

Next up:

- Continued development and refinement
- Code review and cleanup
- Documentation and testing
- Preparing for community feedback

Happy to discuss the approach or answer any questions about the implementation direction! I'm not really a Go dev, but managing to make some meaningful progress 😄

```sh
21:41:15 ~/Documents/Code/rclone icloudphotos ❯ go run . lsd test-icp:
          -1 2025-08-05 21:41:34        -1 AppLibrarySync-com.apple.GenerativePlayground-00000000-0000-0000-0000-000000000001
          -1 2025-08-05 21:41:34        -1 PrimarySync
          -1 2025-08-05 21:41:34        -1 SharedSync-AEAD218B-993F-4EC1-A586-438E791FD976
21:41:34 ~/Documents/Code/rclone icloudphotos ❯ go run . lsd test-icp:PrimarySync
          -1 2025-08-05 21:41:45        -1 All Photos
          -1 2025-08-05 21:41:45        -1 Bursts
          -1 2025-08-05 21:41:45        -1 Favoritas
          -1 2025-08-05 21:41:45        -1 Favorites
          -1 2025-08-05 21:41:45        -1 Hidden
          -1 2025-08-05 21:41:45        -1 Live
          -1 2025-08-05 21:41:45        -1 Panoramas
          -1 2025-08-05 21:41:45        -1 Recently Deleted
          -1 2025-08-05 21:41:45        -1 Screenshots
          -1 2025-08-05 21:41:45        -1 Slo-mo
          -1 2025-08-05 21:41:45        -1 Tech
          -1 2025-08-05 21:41:45        -1 Time-lapse
          -1 2025-08-05 21:41:45        -1 Videos
          -1 2025-08-05 21:41:45        -1 WhatsApp
21:42:36 ~/Documents/Code/rclone icloudphotos ❯ go run . ls test-icp:PrimarySync/Videos 
  5622065 201b836e-da22-4978-b76e-763f221e2add.mp4
  2948154 3a0aad82-36a7-4e5b-a177-b45ca38c64a8.mp4
  2773712 5bee4ed6-3d37-480d-b8f6-f2b213bb4d27.mp4
  3653670 75972959-5f32-4e04-8c7f-03450db6cf19.mp4
  2476201 8548f104-0c69-4075-98e8-b7b88d45bf96.mp4
  3377791 93d8560a-e015-4526-9eeb-103cf302b650.mp4
  4236578 B094C8CA-C225-44CF-A866-8C87C93B5157.mp4
 15818514 Doorbell_Mar-12-2025-at-10-36-30 AM-GMT.mp4
 11337026 Doorbell_Mar-12-2025-at-10-42-01 AM-GMT.mp4
 12544049 Doorbell_Nov-22-2024-at-6-25-36 PM-GMT.mp4
  8910838 FC64A176-27F8-48ED-977E-BA5CF740C0F6.mp4
21:43:39 ~/Documents/Code/rclone icloudphotos ❯ go run . copy test-icp:PrimarySync/Videos /tmp/backup --progress
Transferred:      706.430 MiB / 22.405 GiB, 3%, 41.654 MiB/s, ETA 8m53s
Checks:                 0 / 0, -, Listed 161
Transferred:           19 / 157, 12%
Elapsed time:        15.9s
Transferring:
 *                                  IMG_0321.MOV: 50% /247.758Mi, 17.571Mi/s, 7s
 *                                  IMG_0421.MOV: 49% /243.476Mi, 15.999Mi/s, 7s
 *                                  IMG_0631.MOV: 37% /316.269Mi, 28.031Mi/s, 7s
 *                                  IMG_0632.MOV: 50% /303.322Mi, 32.156Mi/s, 4s^Cexit status 130
```

**Comment by coughlanio** [comment]
Hey all 👋

I've been working on implementing an iCloud Photos backend for rclone and have made good progress on several key areas:

- Authentication and session handling with iCloud services (leveraging some of @lostb1t's iclouddrive backend code 🎉)
- Core directory/download functionality working and tested
- Support for both PrimarySync and SharedSync libraries
- Read-only backend implementation following rclone conventions

The implementation is showing promising results across different testing scenarios (120k+ photos), including various album types and media formats.

Next up:

- Continued development and refinement
- Code review and cleanup
- Documentation and testing
- Preparing for community feedback

Happy to discuss the approach or answer any questions about the implementation direction! I'm not really a Go dev, but managing to make some meaningful progress 😄

Example (obfuscated and truncated for brevity):

```sh
21:41:15 ~/Documents/Code/rclone icloudphotos ❯ go run . lsd test-icp:
          -1 2025-08-05 21:41:34        -1 AppLibrarySync-com.apple.GenerativePlayground-<uuid>
          -1 2025-08-05 21:41:34        -1 PrimarySync
          -1 2025-08-05 21:41:34        -1 SharedSync-<uuid>
21:41:34 ~/Documents/Code/rclone icloudphotos ❯ go run . lsd test-icp:PrimarySync
          -1 2025-08-05 21:41:45        -1 All Photos
          -1 2025-08-05 21:41:45        -1 Bursts
          -1 2025-08-05 21:41:45        -1 Favoritas
          -1 2025-08-05 21:41:45        -1 Favorites
          -1 2025-08-05 21:41:45        -1 Hidden
          -1 2025-08-05 21:41:45        -1 Live
          -1 2025-08-05 21:41:45        -1 Panoramas
          -1 2025-08-05 21:41:45        -1 Recently Deleted
          -1 2025-08-05 21:41:45        -1 Screenshots
          -1 2025-08-05 21:41:45        -1 Slo-mo
          -1 2025-08-05 21:41:45        -1 Tech
          -1 2025-08-05 21:41:45        -1 Time-lapse
          -1 2025-08-05 21:41:45        -1 Videos
          -1 2025-08-05 21:41:45        -1 WhatsApp
21:42:36 ~/Documents/Code/rclone icloudphotos ❯ go run . ls test-icp:PrimarySync/Videos 
  5622065 <uuid>.mp4
  2948154 <uuid>.mp4
  2773712 <uuid>.mp4
  3653670 <uuid>.mp4
  2476201 <uuid>.mp4
  3377791 <uuid>.mp4
  4236578 <uuid>.mp4
 15818514 <uuid>.mp4
 11337026 <uuid>.mp4
 12544049 <uuid>.mp4
  8910838 <uuid>.mp4
21:43:39 ~/Documents/Code/rclone icloudphotos ❯ go run . copy test-icp:PrimarySync/Videos /tmp/backup --progress
Transferred:      706.430 MiB / 22.405 GiB, 3%, 41.654 MiB/s, ETA 8m53s
Checks:                 0 / 0, -, Listed 161
Transferred:           19 / 157, 12%
Elapsed time:        15.9s
Transferring:
 *                                  IMG_0321.MOV: 50% /247.758Mi, 17.571Mi/s, 7s
 *                                  IMG_0421.MOV: 49% /243.476Mi, 15.999Mi/s, 7s
 *                                  IMG_0631.MOV: 37% /316.269Mi, 28.031Mi/s, 7s
 *                                  IMG_0632.MOV: 50% /303.322Mi, 32.156Mi/s, 4s
```

**Comment by samuelexferri** [comment]
I'm sharing here [kovaacs](https://github.com/kovaacs)' message about the implementation of iCloud Advanced Data Protection.
If you're interested in implementing it, you might find it valuable:

https://github.com/rclone/rclone/issues/8310#issuecomment-2927557265

**Comment by coughlanio** [comment]
@samuelexferri Thanks! I suspect we might want to to share the iCloud connection logic between both `iclouddrive` and `icloudphotos` backends as it's basically 1:1 the same right now, and that's probably an appropriate place for that logic to live. 

I'm probably not going to increase the scope of my current PoC, if we can get both Personal and Shared libraries working, that'll probably solve for 99% of users.

As for code sharing, I don't know what the best practice is for `rclone`? I can probably just import the `iclouddrive` code in `icloudphotos` but that seems to go against the spirit of the backend design?

**Comment by stv0g** [comment]
@coughlanio I am really interested in this, and would like to support you in getting this merged. Let me know if I can review or test this feature. Do you have a public branch with these changes?

**Comment by coughlanio** [comment]
@stv0g I've created a draft PR here: https://github.com/rclone/rclone/pull/8734

**Comment by samuelexferri** [comment]
Hi @coughlanio! I’ve added iCloud Private Cloud Storage (PCS) / Advanced Data Protection (ADP) support in this PR: https://github.com/rclone/rclone/pull/8742

If this proves stable, we can likely reuse the same PCS/ADP flow for the iCloud Photos backend as well in the future.

**Comment by kimsvane** [comment]
Any update on this matter ? 
Would really appreciate the drive & photos download!

**Comment by Geremech20261993** [comment]
​"I am a Python developer and I would like to work on adding the iCloud Photos backend support. I have experience with secure authentication and data structures."

**Comment by clouedoc** [comment]
I am adding 50$ to the bounty for Apple iCloud Photos downloading with ADP enabled.

**Comment by clouedoc** [comment]
Here is a list of bounties with the date.

Maybe the current total amount could be added to the title of the issue?

| Username | Message | Amount | Date |
|---|---|---|---|
| [stv0g](https://github.com/stv0g) | "I am willing to sponsor this feature with a contribution of 50 USD." ([source](https://github.com/rclone/rclone/issues/7982#issuecomment-2262417364)) | $50 | 2024-08-01 |
| [Scripter17](https://github.com/Scripter17) | "offered to extend his funding to 200 USD (100 USD for Photos support)" ([source](https://github.com/rclone/rclone/pull/7717#issuecomment-2261025756)) | $100 | 2024-08-01 |
| [lifenautjoe](https://github.com/lifenautjoe) | "I'd like to extend the funding by $500 USD, ($700 in total)" ([source](https://github.com/rclone/rclone/issues/7982#issuecomment-2747785325)) | $500 | 2025-03-24 |
| [poolpOrg](https://github.com/poolpOrg) | "I'll extend the funding by $1000 :-)" ([source](https://github.com/rclone/rclone/issues/7982#issuecomment-2812130770)) | $1,000 | 2025-04-17 |
| [clouedoc](https://github.com/clouedoc) | "I am adding 50$ to the bounty for Apple iCloud Photos downloading with ADP enabled." ([source](https://github.com/rclone/rclone/issues/7982#issuecomment-4004258274)) | $50 | 2026-03-05 |
| | **Total** | **$1,700** | |

(Sorry for the double-posting, by the way.)

**Comment by suraj-mandal** [comment]
If this feature is still not implemented. I would like to give it a try.

**Comment by xz-dev** [comment]
> I am adding 50$ to the bounty for Apple iCloud Photos downloading with ADP enabled.

Just in case no one find that https://pypi.org/project/pyicloud/ it support photo

**Comment by suraj-mandal** [comment]
Working on it and trying to follow the PR raised beforehand for some references

**Comment by Lyapsus** [comment]
@ncw - and anyone interested really - could you take a look at #9234? There is reading support tested well at least on my machine/iCloud and also writing support is WIP with almost everything already working and tested - but as another PR probably

**Comment by stv0g** [comment]
@clouedoc Thanks for creating the table. I see #9234 as the best candidate to get iCloud photos support in rclone :) Once this is merged, I am very happy wire my bounty to @clouedoc

**Comment by Lyapsus** [comment]
@clouedoc, I would like to invite you to #9234 to test my take at ADP support. Would be glad to see some logs with -vv. Thank you in advance

**Comment by Lyapsus** [comment]
@Scripter17, @lifenautjoe, @poolpOrg - I would like to invite you to test #9234 too and give me your verdict. Thanks!

**Comment by clouedoc** [comment]
Great!

I think the next step is to figure out the bounty distribution.

I will contact the people involved privately and report back with a suggested % distribution.

**Comment by Scripter17** [comment]
I physically have the cash on me I just need to go to the bank later today (maybe tomorrow) to deposit it then I can send it over

**Comment by alexanmtz** [comment]
I really wonder why you just pay on Gitpay where this issue was already imported (https://gitpay.me/#/task/1113) and the contributor who solved can claim the bounty and get it all in his bank account? Instead of you all manually sent or even handle this manually here :p

**Comment by clouedoc** [comment]
> I really wonder why you just pay on Gitpay

Okay, I will help you understand.

- ~~8%~~ 15.4% fee
- no added value at this stage
- repeated advertising for your personal gain in a volunteer-based space

Let me know if you have any more questions 🙂

EDIT: there's also 8% fee for payout in case that isn't clear as well. 

So let's do a little example with my personal bounty. 

1. I wire 50$ to gitpay.
2. The bounty increases by 46 USD (8% fee)
3. Somebody claims the bounty
4. They receive 42.3$ (8% fee)

You're suggesting us to pay a 15.4% fee for no service.

**Comment by clouedoc** [comment]
Still waiting on an answer to the email I sent to @coughlanio and @mikegillan.

**Comment by stv0g** [comment]
@Lyapsus Could you sent me your Paypal address or SEPA bank details for wiring you your bounty?

My mail is stv0g@0l.de

**Comment by Scripter17** [comment]
Preferably just post details here

I'm also not clear on distribution

**Comment by Lyapsus** [comment]
@stv0g just did via email, thank you.
@Scripter17, hello, I'll do the same if you don't mind

**Comment by Scripter17** [comment]
Confirming now that I have sent €86 (just over $100 USD) to IakovTill on paypal.

It's been a pleasure

Edit: Wires got crossed; It was euros not pounds

**Comment by Lyapsus** [comment]
I confirm @Scripter17

**Comment by Lyapsus** [comment]
And @stv0g

**Comment by Lyapsus** [comment]
@lifenautjoe @poolpOrg just checking in on the bounty pledges now that #9234 has merged and this issue is closed

**Comment by poolpOrg** [comment]
hello, sorry @Lyapsus I didn't pay attention to the Github notifications until you mailed me.

Just came back here to double-check that it was a legit mail ;-)

I'll get back in touch by mail to see how we can arrange the payout, cheers

**Comment by lifenautjoe** [comment]
Hey @Lyapsus ! Thanks for the great work! Apologies, Github notifications are not something I keep a look out for. 

Saw you reached out on LinkedIn so lets arrange this over there!

**Comment by Lyapsus** [comment]
Meanwhile I confirm @poolpOrg and @clouedoc
